protocols/shell-geometry-tracker-v1.xml

```diff
@@ -0,0 +1,260 @@
+    </evnt>
+
+    <request name="track_toplevels">
+      <description summary="begins tracking toplevels">
+        This request asks the compositor to track all toplevels created by the
+        given toplevel list that enter the tracking area.
+
+        Upon request, all toplevels that qualify as members of the tracker
+        must be provided via an 'enter' event.
+
+        Attempting to set the toplevel list more than once is an error, as
+        is destroying the toplevel list prior to the destruction of
+        this object.
+      </description>
+
+      <arg name="list" type="object" interface="ext_foreign_toplevel_list_v1"/>
+    </request>
+  </interface>
+
+  <interface name="shell_foreign_toplevel_geometry_tracker_member_v1" version="1">
+    <description summary="tracked entity visible on a geometry tracker">
+      A toplevel tracked by a geometry tracker.
+
+      Upon creation, toplevel, geometry, and optionally stacking order
+      events must be sent by the compositor.
+    </description>
+
+    <request name="destroy" type="destructor">
+      <description summary="destroy the member">
+        Destroys the tracker member object.
+      </description>
+    </request>
+
+    <event name="left">
+      <description summary="the toplevel has left the tracked area">
+        The toplevel has left the tracked area. A toplevel must not be
+        considered to have left until the toplevel's geometry has no
+        intersection with the geometry tracker.
+
+        Following this event, the member object should be destroyed and no more
+        events will be sent to it. If the toplevel later re-enters the tracked
+        area it will do so as a new object.
+
+        This event is double-buffered and must be followed by a
+        'shell_foreign_toplevel_geometry_tracker_manager_v1.done' event.
+      </description>
+    </event>
+
+    <event name="geometry">
+      <description summary="geometry of the toplevel has changed">
+        New toplevel geometry information is available.
+        This event is sent upon creation of the member object or changes
+        to the toplevel's geometry.
+
+        Reported geometry should include window decorations such as borders
+        but not shadows.
+
+        This event is double-buffered and must be followed by a
+        'shell_foreign_toplevel_geometry_tracker_manager_v1.done' event.
+      </description>
+
+      <arg name="x" type="int"/>
+      <arg name="y" type="int"/>
+      <arg name="width" type="int"/>
+      <arg name="height" type="int"/>
+    </event>
+
+    <event name="above">
+      <description summary="the toplevel is now above another">
+        This event reports that the toplevel is now directly above the
+        given sibling in Z-order.
+
+        This event is double-buffered and must be followed by a
+        'shell_foreign_toplevel_geometry_tracker_manager_v1.done' event.
```

Comment:
wouldn't it be better to do a css-style z-index?

Reply:
I think this might be easier to manage. Most actions that change z index move one window relative to others, e.g. activate, tile/untile, which should end up as a single event to a single toplevel. With indexes the compositor has to tell the client how to shuffle the list, which is more work for both parties.